warehouse/query-core/src/main/java/datawave/query/data/parsers/ShardIndexKey.java

```diff
@@ -0,0 +1,98 @@
+package datawave.query.data.parsers;
+
+import org.apache.accumulo.core.data.ByteSequence;
+import org.apache.accumulo.core.data.Key;
+
+public class ShardIndexKey implements KeyParser {
+
+    private Key key;
+
+    private ByteSequence cq;
+    private int cqSplit;
+
+    private String field;
+    private String value;
+    private String datatype;
+    private String shard;
+
+    @Override
+    public void parse(Key k) {
+        clearState();
+        this.key = k;
+    }
+
+    @Override
+    public void clearState() {
+        this.cq = null;
+        this.cqSplit = -1;
+
+        this.field = null;
+        this.value = null;
+        this.datatype = null;
+        this.shard = null;
+    }
+
+    @Override
+    public String getDatatype() {
+        if (datatype == null) {
+            if (cq == null) {
+                cq = key.getColumnQualifierData();
+                for (int i = cq.length() - 1; i > 0; i--) {
+                    if (cq.byteAt(i) == 0x00) {
+                        cqSplit = i;
+                        break;
+                    }
+                }
+            }
+            datatype = cq.subSequence(cqSplit + 1, cq.length()).toString();
+        }
+        return datatype;
+    }
+
+    public String getShard() {
+        if (shard == null) {
+            if (cq == null) {
+                cq = key.getColumnQualifierData();
+                for (int i = 0; i < cq.length(); i++) {
+                    if (cq.byteAt(i) == 0x00) {
+                        cqSplit = i;
+                        break;
+                    }
+                }
+            }
+            shard = cq.subSequence(0, cqSplit).toString();
```

Comment:
is it worth setting up the datatype here as well?

Reply:
KeyParsers are lazily parsed, omitting the datatype is by design.